File: src/scanners/CemuSource.js
const fsp = require("fs/promises");
const YAML = require("yaml");
const path = require("path");
const fs = require("fs");

const GameDir = require("./GameDir.js");
const convertPath = require("../utils/convertPathPlatform.js");
const config = require("../utils/configFormats.js");

const WiiUEmulationSource = require("./WiiUEmulationSource.js");
const CemuGame = require("../games/CemuGame.js");

// TODO Remove dependency on lutris game
class CemuSource extends WiiUEmulationSource {

	static name = "Cemu in Lutris";
	static gameClass = CemuGame;

	preferCache = false;

	cemuLutrisGame = undefined;
	romRegex = /.+\.(wud|wux|wad|iso|rpx|elf)/i;

	constructor(cemuLutrisGame, preferCache = false) {
		super();
		this.cemuLutrisGame = cemuLutrisGame;
		this.preferCache = preferCache;
	}

	/**
	 * Get cemu's config data
	 * @param {string} cemuExePath - The path to cemu's executable
	 * @returns {object} - Cemu's config data
	 * @private
	 */
	async _getConfig(cemuExePath) {
		const configDir = path.dirname(cemuExePath);
		const configFilePath = `${configDir}/settings.xml`;
		const configFileContents = await fsp.readFile(configFilePath, "utf-8");
		const configData = await config.xml2js(configFileContents);
		return configData;
	}

	/**
	 * Get cemu's cached games from its config data
	 * @param {object} configData - Cemu's config data
	 * @param {string} prefix - Cemu's wine prefix absolute path
	 * @returns {CemuGame[]} - An array of found cached games
	 * @private
	 */
	async _getCachedROMs(configData, prefix) {

		// Search into config for cached games
		const games = [];

		const gameCache = configData?.content?.GameCache?.Entry;
		if (typeof gameCache === "undefined") {
			return;
		}
		for (const item of gameCache) {

			// Get game name
			const customName = item?.custom_name?.[0];
			const defaultName = item?.name?.[0];
			let name = undefined;
			for (let candidate of [customName, defaultName]) {
				if (typeof candidate !== "string") {
					continue;
				}
				candidate = candidate.trim();
				if (candidate.length > 0) {
					name = candidate;
					break;
				}
			}
			if (!name) {
				continue;
			}

			// Get game path
			const winePath = item?.path;
			if (!winePath) {
				continue;
			}

			// Build game
			const linuxPath = convertPath.wineToLinux(winePath, prefix);
			const isInstalled = fs.existsSync(linuxPath);
			const game = new this.constructor.gameClass(name, linuxPath);
			game.isInstalled = isInstalled;

			// Get more info
			if (isInstalled && path.extname(linuxPath) === ".rpx") {
				this._getRPXGameProps(game);
			}

			games.push(game);
		}

		return games;

	}

	/**
	 * Get cemu's game dirs from its config data
	 * @param {object} configData - Cemu's config data
	 * @param {string} prefix - Cemu's wine prefix absolute path
	 * @returns {GameDir[]} - The game dirs extracted from cemu's config
	 * @private
	 */
	async _getROMDirs(configData, prefix) {

		// Search into config for ROM dirs
		//const wPaths = configData?.content?.GamePaths?.[0]?.Entry;
		const wPaths = [];
		const entries = configData?.content?.GamePaths?.Entry;
		if (!entries) {
			return [];
		}
		if (Array.isArray(entries)) {
			wPaths.push(...entries);
		} else {
			wPaths.push(entries);
		}
		const lPaths = wPaths.map(wPath=>convertPath.wineToLinux(wPath, prefix));
		const gameDirs = lPaths.map(lPath=>new GameDir(lPath, true));
		return gameDirs;

	}

	/**
	 * Get cemu ROM games from given game directories
	 * @param {GameDir[]} dirs - The game dirs to scan for ROMs
	 * @param {boolean} warn - Whether to display additional warnings
	 * @returns {CemuGame[]} - An array of found games
	 * @private
	 */
	async _getROMGames(dirs, warn = false) {

		// Scan cemu dirs
		const gameRomPaths = await this._getROMs(dirs, this.romRegex, warn);

		// Convert found paths into cemu games
		const romGamesPromises = gameRomPaths.map(async (romPath)=>{

			// Get base info
			const basename = path.basename(romPath);
			const game = new this.constructor.gameClass(basename, romPath);
			game.isInstalled = true;

			// Precise game info
			const extname = path.extname(romPath).toLowerCase();
			if (extname === ".rpx") {
				await this._getRPXGameProps(game);
			}

			return game;
		});

		const romGames = await Promise.all(romGamesPromises);
		return romGames;
	}

	/**
	 * Get all cemu games
	 * @param {boolean} warn - Whether to display additional warnings
	 * @returns {CemuGame[]} - An array of found games
	 */
	async scan(warn = false) {

		// Read lutris config for cemu (to get cemu's exe path)
		const USER_DIR = process.env["HOME"];
		const lutrisConfigPath = path.join(USER_DIR, ".config/lutris/games", `${this.cemuLutrisGame.configPath}.yml`);
		let cemuExePath, cemuPrefixPath;
		try {
			const lutrisConfigContents = await fsp.readFile(lutrisConfigPath, "utf-8");
			const parsedLutrisConfig = YAML.parse(lutrisConfigContents);
			cemuExePath = parsedLutrisConfig.game.exe;
			cemuPrefixPath = parsedLutrisConfig.game.prefix;
		} catch (error) {
			if (warn){
				console.warn(`Unable to read lutris's game config file for cemu : ${error}`);
			}
		}

		// Read cemu's config
		let configData;
		if (typeof cemuExePath !== "undefined") {
			try {
				configData = await this._getConfig(cemuExePath);
			} catch (error) {
				if (warn){
					console.warn(`Unable to read cemu config file : ${error}`);
				}
			}
		}

		// If (scan) : scan cemu's game paths for games and ignore cemu's game cache
		// Else      : trust cemu's game cache
		let romGames = [];
		if (typeof configData !== "undefined" && typeof cemuPrefixPath !== "undefined") {

			if (!this.preferCache) {

				// Get cemu's ROM dirs
				let romDirs = [];
				try {
					romDirs = await this._getROMDirs(configData, cemuPrefixPath);
				} catch (error) {
					if (warn){
						console.warn(`Unable to get cemu ROM dirs : ${error}`);
					}
				}

				// Scan ROMDirs for ROMs
				if (romDirs.length > 0) {
					try {
						romGames = await this._getROMGames(romDirs, warn);
					} catch (error) {
						if (warn){
							console.warn(`Unable to get cemu ROMs : ${error}`);
						}
					}
				}

			} else {

				// Get cemu's cached ROM games
				try {
					romGames = await this._getCachedROMs(configData, cemuPrefixPath);
				} catch (error) {
					if (warn){
						console.warn(`Unable to get cemu cached ROMs : ${error}`);
					}
				}

			}
		}

		return romGames;

	}

}

module.exports = CemuSource;

File: gali/sources/cemu_source.py
import yaml
import sqlite3
from math import inf
from locale import getlocale, LC_MESSAGES
from pathlib import PurePath
from xml.etree.ElementTree import ElementTree

from gali.utils.locations import HOME
from gali.sources.emulation_source import EmulationSource
from gali.utils.rpx_metadata import RPXMetadata
from gali.games.cemu_game import CemuLutrisGame
from gali.sources.lutris_source import LutrisSource
from gali.utils.wine_path import wine_to_posix
from gali.sources.game_dir import GameDir

class CemuLutrisSource(EmulationSource):

	name : str = "Cemu (Lutris)"
	game_class : type[CemuLutrisGame] = CemuLutrisGame
	rom_extensions : tuple[str] = (".wud", ".wux", ".wad", ".iso", ".rpx", ".elf")

	def get_cemu_lutris_config(self) -> dict:

		# Get path to lutris config for cemu
		connection = sqlite3.connect(LutrisSource.db_path)
		sql = "SELECT configpath FROM 'games' WHERE slug = 'cemu'"
		cursor = connection.execute(sql)
		row = cursor.fetchone()
		connection.close()
		config_path = f"{HOME}/.config/lutris/games/{row[0]}.yml"

		# Get config content
		file = open(config_path, "r", encoding="utf-8-sig")
		config = yaml.safe_load(file)
		file.close()
		return config

	def get_cemu_config(self, cemu_exe_path) -> ElementTree:
		config_dir = PurePath(cemu_exe_path).parent
		config_path = f"{config_dir}/settings.xml"
		config = ElementTree()
		config.parse(config_path)
		return config

	def get_cached_games(self, wine_prefix_path: str, config: ElementTree) -> tuple[CemuLutrisGame]:
		
		games = []

		# Read XML tree
		elements = config.findall("./GameCache/Entry")
		for element in elements:

			# Get game data
			name = element.findtext("name", default=None)
			game_path = element.findtext("path", default=None)
			if name is None or game_path is None:
				continue

			# Convert wine path to posix
			game_path = wine_to_posix(wine_prefix_path, game_path)

			# Build game
			game = self.game_class(
				name=name,
				game_path=game_path
			)
			games.append(game)

		return tuple(games)

	def get_rom_dirs(self, wine_prefix_path: str, config : ElementTree) -> tuple[str]:
		game_dirs = []
		elements = config.findall("./GamePaths/Entry")
		for element in elements:
			path = wine_to_posix(wine_prefix_path, element.text)
			print(path) # ! DEBUG
			game_dir = GameDir(path, inf)
			game_dirs.append(game_dir)
		return tuple(game_dirs)

	def get_rom_games(self, rom_dirs : tuple[str]) -> tuple[CemuLutrisGame]:

		games = []

		# Get locale language (for rpx games)
		locale = getlocale(LC_MESSAGES)
		locale_lang_code = locale[0]
		if locale_lang_code is not None:
			locale_lang_code = locale_lang_code[0:2]
			locale_lang_code = locale_lang_code.lower()

		# Scan every rom dir
		for rom_dir in rom_dirs:

			# Scan for roms
			try:
				rom_paths = self.get_rom_paths(rom_dir, self.rom_extensions)
				pass
			except Exception:
				continue

			# Build games from roms
			for rom_path in rom_paths:
				pure_path = PurePath(rom_path)
				extension = pure_path.suffix
				basename = pure_path.stem

				# Special case for metadata-rich ".rpx" games
				if extension == ".rpx":
					metadata : RPXMetadata = RPXMetadata.from_rom_path(rom_path)
					name = metadata.long_name.get(locale_lang_code, basename)
					game = self.game_class(name=name, game_path=rom_path)
					games.append(game)

				# Other games
				else:
					game = self.game_class(name=basename, game_path=rom_path)
					games.append(game)

		return games

	def scan(self) -> list[CemuLutrisGame]:

		# Read lutris config for cemu
		cemu_lutris_config = self.get_cemu_lutris_config()
		wine_prefix_path = cemu_lutris_config["game"]["prefix"]
		exe_path = cemu_lutris_config["game"]["exe"]
		exe_path = f"{wine_prefix_path}/{exe_path}"

		# Read cemu's config
		cemu_config = self.get_cemu_config(exe_path)

		# Scan for games
		if self.prefer_cache:
			games = self.get_cached_games(wine_prefix_path, cemu_config)
		else:
			game_dirs = self.get_rom_dirs(wine_prefix_path, cemu_config)
			games = self.get_rom_games(game_dirs)
		return games